In [1]:
import pandas as pd
import numpy as np

In [2]:
sp_df = pd.read_csv('StudentsPerformance.csv')
sp_df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

In [3]:
mean_scores = sp_df.groupby(['gender', 'lunch']) \
                   .aggregate({'math score': 'mean', 'reading score': 'mean'}) \
                   .rename(columns={'math score': 'mean_math_score', 'reading score': 'mean_reading_score'})

In [4]:
mean_scores

mean_math_score  mean_reading_score
gender lunch                                            
female free/reduced        55.814815           67.386243
       standard            68.124620           75.607903
male   free/reduced        62.457831           61.542169
       standard            72.022152           67.537975

In [5]:
mean_scores.loc[[('female', 'standard'), ('female', 'free/reduced')]] # многоуровневые индексы

mean_math_score  mean_reading_score
gender lunch                                            
female standard            68.124620           75.607903
       free/reduced        55.814815           67.386243

In [6]:
sp_df.groupby(['gender']).lunch.unique()

gender
female    [standard, free/reduced]
male      [free/reduced, standard]
Name: lunch, dtype: object

In [7]:
sp_df.sort_values(['gender', 'math score'], ascending=False).groupby('gender').head(10)

gender race/ethnicity parental level of education         lunch  \
149    male        group E          associate's degree  free/reduced   
623    male        group A                some college      standard   
625    male        group D                some college      standard   
916    male        group E           bachelor's degree      standard   
306    male        group E                some college      standard   
104    male        group C                some college      standard   
934    male        group C          associate's degree      standard   
34     male        group E                some college      standard   
286    male        group E          associate's degree      standard   
539    male        group A          associate's degree      standard   
451  female        group E                some college      standard   
458  female        group E           bachelor's degree      standard   
962  female        group E          associate's degree      standard   
114  female        group E           bachelor's degree      standard   
263  female        group E                 high school      standard   
712  female        group D                some college      standard   
179  female        group D            some high school      standard   
855  female        group B           bachelor's degree      standard   
165  female        group C           bachelor's degree      standard   
717  female        group C          associate's degree      standard   

    test preparation course  math score  reading score  writing score  
149               completed         100            100             93  
623               completed         100             96             86  
625               completed         100             97             99  
916               completed         100            100            100  
306               completed          99             87             81  
104               completed          98             86             90  
934               completed          98             87             90  
34                     none          97             87             82  
286               completed          97             82             88  
539               completed          97             92             86  
451                    none         100             92             97  
458                    none         100            100            100  
962                    none         100            100            100  
114               completed          99            100            100  
263                    none          99             93             90  
712                    none          98            100             99  
179               completed          97            100            100  
855                    none          97             97             96  
165               completed          96            100            100  
717               completed          96             96             99

In [8]:
total_score = sp_df.filter(like='score').sum(axis=1)
sp_df['total score'] = total_score

In [9]:
sp_df

gender race/ethnicity parental level of education         lunch  \
0    female        group B           bachelor's degree      standard   
1    female        group C                some college      standard   
2    female        group B             master's degree      standard   
3      male        group A          associate's degree  free/reduced   
4      male        group C                some college      standard   
..      ...            ...                         ...           ...   
995  female        group E             master's degree      standard   
996    male        group C                 high school  free/reduced   
997  female        group C                 high school  free/reduced   
998  female        group D                some college      standard   
999  female        group D                some college  free/reduced   

    test preparation course  math score  reading score  writing score  \
0                      none          72             72             74   
1                 completed          69             90             88   
2                      none          90             95             93   
3                      none          47             57             44   
4                      none          76             78             75   
..                      ...         ...            ...            ...   
995               completed          88             99             95   
996                    none          62             55             55   
997               completed          59             71             65   
998               completed          68             78             77   
999                    none          77             86             86   

     total score  
0            218  
1            247  
2            278  
3            148  
4            229  
..           ...  
995          282  
996          172  
997          195  
998          223  
999          249  

[1000 rows x 9 columns]

In [10]:
sp_df = sp_df.assign(total_score_log=np.log(sp_df['total score']))

In [11]:
sp_df

gender race/ethnicity parental level of education         lunch  \
0    female        group B           bachelor's degree      standard   
1    female        group C                some college      standard   
2    female        group B             master's degree      standard   
3      male        group A          associate's degree  free/reduced   
4      male        group C                some college      standard   
..      ...            ...                         ...           ...   
995  female        group E             master's degree      standard   
996    male        group C                 high school  free/reduced   
997  female        group C                 high school  free/reduced   
998  female        group D                some college      standard   
999  female        group D                some college  free/reduced   

    test preparation course  math score  reading score  writing score  \
0                      none          72             72             74   
1                 completed          69             90             88   
2                      none          90             95             93   
3                      none          47             57             44   
4                      none          76             78             75   
..                      ...         ...            ...            ...   
995               completed          88             99             95   
996                    none          62             55             55   
997               completed          59             71             65   
998               completed          68             78             77   
999                    none          77             86             86   

     total score  total_score_log  
0            218         5.384495  
1            247         5.509388  
2            278         5.627621  
3            148         4.997212  
4            229         5.433722  
..           ...              ...  
995          282         5.641907  
996          172         5.147494  
997          195         5.273000  
998          223         5.407172  
999          249         5.517453  

[1000 rows x 10 columns]

In [12]:
sp_df.drop(labels=['total score', 'total_score_log'], axis=1)

gender race/ethnicity parental level of education         lunch  \
0    female        group B           bachelor's degree      standard   
1    female        group C                some college      standard   
2    female        group B             master's degree      standard   
3      male        group A          associate's degree  free/reduced   
4      male        group C                some college      standard   
..      ...            ...                         ...           ...   
995  female        group E             master's degree      standard   
996    male        group C                 high school  free/reduced   
997  female        group C                 high school  free/reduced   
998  female        group D                some college      standard   
999  female        group D                some college  free/reduced   

    test preparation course  math score  reading score  writing score  
0                      none          72             72             74  
1                 completed          69             90             88  
2                      none          90             95             93  
3                      none          47             57             44  
4                      none          76             78             75  
..                      ...         ...            ...            ...  
995               completed          88             99             95  
996                    none          62             55             55  
997               completed          59             71             65  
998               completed          68             78             77  
999                    none          77             86             86  

[1000 rows x 8 columns]

## Заполните пропуски

In [13]:
dota_df = pd.read_csv('dota_hero_stats.csv')
dota_df

Unnamed: 0 attack_type   id  legs  localized_name  \
0             0       Melee    1     2       Anti-Mage   
1             1       Melee    2     2             Axe   
2             2      Ranged    3     4            Bane   
3             3       Melee    4     2     Bloodseeker   
4             4      Ranged    5     2  Crystal Maiden   
..          ...         ...  ...   ...             ...   
112         112       Melee  114     2     Monkey King   
113         113      Ranged  119     2     Dark Willow   
114         114       Melee  120     2       Pangolier   
115         115      Ranged  121     2      Grimstroke   
116         116       Melee  129     2            Mars   

                             name primary_attr  \
0          npc_dota_hero_antimage          agi   
1               npc_dota_hero_axe          str   
2              npc_dota_hero_bane          int   
3       npc_dota_hero_bloodseeker          agi   
4    npc_dota_hero_crystal_maiden          int   
..                            ...          ...   
112     npc_dota_hero_monkey_king          agi   
113     npc_dota_hero_dark_willow          int   
114       npc_dota_hero_pangolier          agi   
115      npc_dota_hero_grimstroke          int   
116            npc_dota_hero_mars          str   

                                                 roles  
0                         ['Carry', 'Escape', 'Nuker']  
1      ['Initiator', 'Durable', 'Disabler', 'Jungler']  
2          ['Support', 'Disabler', 'Nuker', 'Durable']  
3    ['Carry', 'Disabler', 'Jungler', 'Nuker', 'Ini...  
4          ['Support', 'Disabler', 'Nuker', 'Jungler']  
..                                                 ...  
112       ['Carry', 'Escape', 'Disabler', 'Initiator']  
113         ['Support', 'Nuker', 'Disabler', 'Escape']  
114  ['Carry', 'Nuker', 'Disabler', 'Durable', 'Esc...  
115         ['Support', 'Nuker', 'Disabler', 'Escape']  
116      ['Carry', 'Initiator', 'Disabler', 'Durable']  

[117 rows x 8 columns]

In [14]:
dota_df.groupby('legs', as_index=False).count()

legs  Unnamed: 0  attack_type  id  localized_name  name  primary_attr  \
0     0          11           11  11              11    11            11   
1     2          95           95  95              95    95            95   
2     4           7            7   7               7     7             7   
3     6           3            3   3               3     3             3   
4     8           1            1   1               1     1             1   

   roles  
0     11  
1     95  
2      7  
3      3  
4      1

## Отметьте верные ячейки

In [15]:
acc_df = pd.read_csv('accountancy.csv')
acc_df.head()

Unnamed: 0 Executor Type  Salary
0           0     Pupa    D      63
1           1     Pupa    A     158
2           2     Pupa    D     194
3           3     Pupa    E     109
4           4    Loopa    E     184

In [16]:
acc_df.groupby(['Type', 'Executor']).aggregate({'Salary': 'max'})

Salary
Type Executor        
A    Loopa         58
     Pupa         234
B    Loopa        249
     Pupa          77
C    Loopa        249
     Pupa          94
D    Loopa        181
     Pupa         194
E    Loopa        220
     Pupa         197
F    Loopa        244
     Pupa         195

## Выберите один вариант из списка

In [17]:
df = pd.read_csv('dota_hero_stats.csv')
df = df.drop('Unnamed: 0', axis=1)
df

attack_type   id  legs  localized_name                          name  \
0         Melee    1     2       Anti-Mage        npc_dota_hero_antimage   
1         Melee    2     2             Axe             npc_dota_hero_axe   
2        Ranged    3     4            Bane            npc_dota_hero_bane   
3         Melee    4     2     Bloodseeker     npc_dota_hero_bloodseeker   
4        Ranged    5     2  Crystal Maiden  npc_dota_hero_crystal_maiden   
..          ...  ...   ...             ...                           ...   
112       Melee  114     2     Monkey King     npc_dota_hero_monkey_king   
113      Ranged  119     2     Dark Willow     npc_dota_hero_dark_willow   
114       Melee  120     2       Pangolier       npc_dota_hero_pangolier   
115      Ranged  121     2      Grimstroke      npc_dota_hero_grimstroke   
116       Melee  129     2            Mars            npc_dota_hero_mars   

    primary_attr                                              roles  
0            agi                       ['Carry', 'Escape', 'Nuker']  
1            str    ['Initiator', 'Durable', 'Disabler', 'Jungler']  
2            int        ['Support', 'Disabler', 'Nuker', 'Durable']  
3            agi  ['Carry', 'Disabler', 'Jungler', 'Nuker', 'Ini...  
4            int        ['Support', 'Disabler', 'Nuker', 'Jungler']  
..           ...                                                ...  
112          agi       ['Carry', 'Escape', 'Disabler', 'Initiator']  
113          int         ['Support', 'Nuker', 'Disabler', 'Escape']  
114          agi  ['Carry', 'Nuker', 'Disabler', 'Durable', 'Esc...  
115          int         ['Support', 'Nuker', 'Disabler', 'Escape']  
116          str      ['Carry', 'Initiator', 'Disabler', 'Durable']  

[117 rows x 7 columns]

In [18]:
df.groupby(['attack_type', 'primary_attr']).count().id

attack_type  primary_attr
Melee        agi             19
             int              2
             str             35
Ranged       agi             18
             int             40
             str              3
Name: id, dtype: int64

## Напишите программу. Тестируется через stdin → stdout

In [20]:
algae_df = pd.read_csv('algae.csv')
algae_df

species        genus  group   sucrose    alanin   citrate  \
0     Fucus_vesiculosus        Fucus  brown  3.001472  3.711498  5.004262   
1   Saccharina_japonica   Saccharina  brown  6.731070  1.255251  5.621499   
2        Fucus_serratus        Fucus  brown  3.276870  0.346431  1.216767   
3       Fucus_distichus        Fucus  brown  6.786996  6.641303  6.423606   
4     Cladophora_fracta   Cladophora  green  3.861470  1.648450  6.940588   
5   Cladophora_compacta   Cladophora  green  5.712284  3.461692  3.082826   
6   Cladophora_gracilis   Cladophora  green  2.452623  6.881024  5.841520   
7      Palmaria_palmata     Palmaria    red  0.704580  3.176440  5.573905   
8  Saccharina_latissima   Saccharina  brown  1.636122  5.793163  1.071920   
9   Ascophyllum_nodosum  Ascophyllum  brown  6.825467  0.875429  5.253527   

    glucose  oleic_acid  
0  2.548459    6.405165  
1  6.013219    4.156700  
2  3.623225    0.304573  
3  2.272724    3.393203  
4  2.316955    2.528886  
5  3.343707    1.432514  
6  2.740165    2.829016  
7  3.242090    2.245538  
8  3.947968    4.817804  
9  3.414961    2.432526

In [31]:
algae_df.groupby(['genus']).mean()

sucrose    alanin   citrate   glucose  oleic_acid
genus                                                          
Ascophyllum  6.825467  0.875429  5.253527  3.414961    2.432526
Cladophora   4.008792  3.997055  5.288311  2.800276    2.263472
Fucus        4.355112  3.566411  4.214878  2.814803    3.367647
Palmaria     0.704580  3.176440  5.573905  3.242090    2.245538
Saccharina   4.183596  3.524207  3.346710  4.980594    4.487252

## Напишите текст

In [35]:
algae_df.query('genus == "Fucus"').agg(['mean', 'min', 'max']).alanin

min     0.346431
max     6.641303
mean    3.566411
Name: alanin, dtype: float64

## Сопоставьте значения из двух списков
Размах (максимум - минимум) сахарозы в бурых, зелёных и красных водорослях - 5.19, 3.26, 0

Дисперсия концентрации цитрата в бурых, зелёных и красных водорослях - 5.47, 3.95, NaN

Число видов бурых, зелёных и красных водорослей - 6, 3, 1

In [39]:
algae_df.groupby(['group']).describe()

sucrose                                                             \
        count      mean       std       min       25%      50%       75%   
group                                                                      
brown     6.0  4.709666  2.336471  1.636122  3.070321  5.00397  6.773014   
green     3.0  4.008792  1.634817  2.452623  3.157047  3.86147  4.786877   
red       1.0  0.704580       NaN  0.704580  0.704580  0.70458  0.704580   

                alanin            ...   glucose           oleic_acid  \
            max  count      mean  ...       75%       max      count   
group                             ...                                  
brown  6.825467    6.0  3.103846  ...  3.866782  6.013219        6.0   
green  5.712284    3.0  3.997055  ...  3.041936  3.343707        3.0   
red    0.704580    1.0  3.176440  ...  3.242090  3.242090        1.0   

                                                                             
           mean       std       min       25%       50%       75%       max  
group                                                                        
brown  3.584995  2.094095  0.304573  2.672695  3.774952  4.652528  6.405165  
green  2.263472  0.735111  1.432514  1.980700  2.528886  2.678951  2.829016  
red    2.245538       NaN  2.245538  2.245538  2.245538  2.245538  2.245538  

[3 rows x 40 columns]

In [43]:
algae_df.groupby('group') \
        .agg({'citrate': 'var', 'sucrose': lambda x: x.max() - x.min(), 'glucose': 'count'}) \
        .round(2)

citrate  sucrose  glucose
group                           
brown     5.47     5.19        6
green     3.95     3.26        3
red        NaN     0.00        1